File: src/main/java/com/example/programmercarl/chain_table/reverse_linked_list/Improve.java
package com.example.programmercarl.chain_table.reverse_linked_list;

import java.util.Scanner;

public class Improve {

    public static ListNode reverseList(ListNode head){
            ListNode l, r, temp;
            l = null;
            temp = head;
            r = null;
            while (temp.next != null){
                r = temp.next;
                temp.next = l;
                l = temp;
                temp = r;
            }
            return l;
    }
    public static void main(String[] args) {
        ListNode head = new ListNode();
        ListNode p = head;
        Scanner sc = new Scanner(System.in);
        System.out.println("请输入链表的元素");
        String str = sc.nextLine();
        String[] arr = str.split(",");
        int[] num = new int[arr.length];
        for(int i = 0; i < arr.length; i++){
            num[i] = Integer.parseInt(arr[i]);
            ListNode node = new ListNode(num[i]);
            p.next = node;
            node.next = null;
            p = p.next;
        }
        p.next = null;
        p = head;
        while(p != null){
            System.out.print(p.val + " ");
            p = p.next;
        }
        System.out.println();
//        来个左右指针法

        p = reverseList(head);
        while(p != null){
            System.out.print(p.val + " ");
            p = p.next;
        }


    }
}
//结果：
/*
执行用时分布
0
ms
击败
100.00%
复杂度分析
消耗内存分布
41.33
MB
击败
77.84%

 */

File: src/main/java/com/example/hot100/linked_list/has_cycle/improve.js
//快慢指针法
//想象兔子和乌龟在同一跑道上，一个速度快、另一个速度慢。如果跑道有环，
// 兔子必然在一段时间后追上乌龟。对于链表来说，如果在链表中引入两个以不同速度（一个比另一个快一倍）
// 前进的指针，在链表存在环的情况下，这两个指针必定会相遇。

let hasCycle = function (head) {
    //定义快慢指针,当快慢指针相遇则说明有环
    let fast = head;
    let slow = head;
    while (fast && fast.next) {
        fast = fast.next.next;
        slow = slow.next;
        if(fast === slow) {
            return true;
        }
    }
    return false;

};

//执行用时分布
// 55
// ms
// 击败
// 90.46%
// 复杂度分析
// 消耗内存分布
// 59.69
// MB
// 击败
// 14.42%